[Work in progress] Investigate using selective/local imports.

SamAxe opened the pull request:
[WIP] Please do not merge this as is.

This is a follow up to the dtemplate.d pull request
(https://github.com/dlang/dmd/pull/11799), and (https://github.com/dlang/dmd/pull/11806), I'm trying a draft PR this time to see if that does what I should have been doing.. 

I'm looking for feedback from CI and any other
comments. My intention is that this is an intermediate step and the
symbols that are only referenced 1 or 2 times will be made local imports.

Other comments:

This patch was tested on Ubuntu 20.04. No windows tests.
There are obvious formatting issues that I would address in a subsequent
resubmission.
The imports that effectively import the entire public interface
shouldn't use selective imports.
This is a very large changeset. No idea what the strategy should be
for applying it if it's desirable.
Some steps for consideration
1) remove unused imports,
2) convert to selective imports,
3) convert to local imports
4) Do 1-3 on a file by file basis?

Edit: tested with dmd-2.091.1 in addition to dmd-2.093.1
P.S. Also going to try submitting pull request with other option to
see if it will just build without requesting a needless review from the
world.

SamAxe committed:
[work in progress] Attempting to remove unused imports

nordlow commented:
I'm working diagnostics at https://github.com/dlang/dmd/pull/11600 that detects, among other things, unused local imports. I'm also trying to figure out if there's a way to detect which symbols for a normal import that are used and propose to convert that import to a (set) of local imports.

thewilsonator commented:
Please split this PR into one containing the removal of unused imports. Note that imports of `core.stdc.stdio` are there for the various diagnostics so leave them be.

SamAxe commented:
@thewilsonator Here is a rebased version with unused imports commented out.  I can update this one with the lines removed.  What else do I need to tidy up for this first PR?  Thanks.

thewilsonator commented:
> unused imports commented out

Please remove the unused imports, I was expecting to see a lot more red in the diff.

thewilsonator commented:
(Sorry hit the wrong button.)

SamAxe committed:
Remove unused imports.

Tested with
1) ./build.d
2) ./build.d unittest
3) ./build.d html

on Ubuntu 20.04, dmd-2.094.0

This was determined by a brute force comment the import out and see if
the recompile was successful.  It would be much nicer if the compiler
could report referenced symbols/imports.

SamAxe commented:
This PR removes unused imports that resulted from looking into using selective and local imports which may be explored in subsequent PRs.

SamAxe commented:
> Please remove the unused imports, I was expecting to see a lot more red in the diff.

There are a number of cases where imports are using 1 symbol in 1 place.  Maybe that is a natural next step from this patch.

SamAxe commented:
@thewilsonator This patch should be just the removed imports.  I don't know how the PR process works.  Is this sufficient for you to take it from here or do I need to open a new PR to better title what this ended up as, etc.  Thanks.

ibuclaw commented:
FYI, you should test any refactoring of imports by compiling all modules one at a time and seeing if dmd still links.

A regression has already made it into master from made by a shuffling of imports.

thewilsonator commented on src/dmd/arrayop.d:
```diff
@@ -13,6 +13,7 @@
 
 module dmd.arrayop;
 
+// import core.stdc.stdio : ;
 import core.stdc.stdio;
```
What I mean is if `core.stdc.stdio` is unused in the given file this diff should read 
```diff
-import core.stdc.stdio;
```

SamAxe replied:
I misinterpreted that the stdio was programmer preference rather than caused build failures and I flubbed up some of the patch. :/

But the more I dig into imports and the fragility of the code base, I'm not familiar enough with the nuances of D semantics to be effective here.  I'll delete this fork in a couple of days and would recommend the patch was reverted in master if it hasn't already been done.  I don't understand the implications of changing imports and the compiler isn't strict enough to protect me here either.

Some of the issues that I found that I think should be addressed before import modification:
1)  Making an import selective and/or local can double the compile time or significantly increase it
2)  Reordering imports causes symbols to not be public in other modules.
3)  Reordering imports can cause template recursion errors
4)  Apparently there are linking issues, but I did not observe these.

I did observe localizing imports for a few key modules did decrease compile time by 10-20%, so seem like it's worth doing.

Thank you for your time and sorry for the distractions.
